[HUDI-1996] Adding functionality to allow the providing of basic auth creds for confluent cloud schema registry

jhsb25 opened the pull request:
## *Tips*
- *Thank you very much for contributing to Apache Hudi.*
- *Please review https://hudi.apache.org/contributing.html before opening a pull request.*

## What is the purpose of the pull request

Ticket - https://issues.apache.org/jira/browse/HUDI-1996

The purpose of this pull request is to allow connectivity with confluent schema registries that require basic authentication (which confluent cloud requires). It does this by allowing the user to provide username and password in the `hoodie.deltastreamer.schemaprovider.registry.baseUrl` so that it is available to the `SchemaRegistryProvider` and by upgrading the version of the io.confluent schema registry libs that are bundled up with hudi as the current ones included do not support basic auth. Despite this issue https://github.com/apache/hudi/issues/1925 being closed the error still persists due to a JDK described here https://bugs.openjdk.java.net/browse/JDK-5043482.  This is pretty high priority for us as without these changes using hudi with confluent cloud is pretty much impossible without custom code changes.


## Brief change log

  - *Bumping io.confluent libs bundled to support basic auth with schema registry.*
  - *Allowing the passing of basic auth in url for hudi `SchemaRegistryProvider`*

## Verify this pull request

*(Please pick either of the following options)*

This change added tests and can be verified as follows:

  - This PR includes unit test for the changes made in the file .
  - Regression tests covered by existing tests.

## Committer checklist

 - [x] Has a corresponding JIRA in PR title & commit
 
 - [x] Commit message is descriptive of the change
 
 - [x] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

n3nash commented:
@nsivabalan Can you review this please ?

nsivabalan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/schema/SchemaRegistryProviderTest.java:
```diff
@@ -0,0 +1,116 @@
+
+  @Test
+  public void testGetTargetSchemaShouldRequestSchemaWithCreds() throws IOException {
+    InputStream is = new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8));
```
Can we re-use code across previous test and this one. except for 1 or 2 lines, mostly its same. we can create a private method w/ an argument for whether its source schema or target schema

jhsb25 replied:
done

nsivabalan commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/schema/SchemaRegistryProviderTest.java:
```diff
@@ -0,0 +1,116 @@
+  @Test
+  public void testGetTargetSchemaShouldRequestSchemaWithoutCreds() throws IOException {
+    TypedProperties props = getProps();
+    props.put("hoodie.deltastreamer.schemaprovider.registry.url", "http://localhost");
```
same comment as above.

jhsb25 replied:
done

nsivabalan reviewed (commented):
Thanks for your contribution. Seems a valuable addition. 
Few clarification: 
I see that you have mentioned you have to make changes to your forked hudi to get this done. Is it just about overriding or adding your own SchemaProvider or you mean to refer to the confluent artifact versions being used? I mean, if we upgrade our confluent artifact versions w/o any changes to schema registry, would you be able to provide your own schema provider registry and get your flow working? Just trying to understand the changes required in general. 

Also, can you confirm if you have tested this patch in your staging env for both w/ and w/o credentials and it all succeeded. 

@codope : I remember you had a local confluent set up right. Do you think you can test this patch out. Also, feel free to review the patch.

codope commented:
> @codope : I remember you had a local confluent set up right. Do you think you can test this patch out. Also, feel free to review the patch.

I tested this patch. It works. The current master does not work with basic auth.

codope commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/schema/SchemaRegistryProvider.java:
```diff
@@ -48,19 +54,40 @@
     return node.get("schema").asText();
   }
 
+  public void setAuthorizationHeader(String creds, HttpURLConnection connection) {
```
Probably because it's being used in tests. If that's the only reason then maybe we can make the method package-private?

jhsb25 replied:
yes thats right, this is implemented like this for testing purposes in order to allow for mocking/spying of the method. I've made the method package-private.

jhsb25 commented:
> Thanks for your contribution. Seems a valuable addition.
> Few clarification:
> I see that you have mentioned you have to make changes to your forked hudi to get this done. Is it just about overriding or adding your own SchemaProvider or you mean to refer to the confluent artifact versions being used? I mean, if we upgrade our confluent artifact versions w/o any changes to schema registry, would you be able to provide your own schema provider registry and get your flow working? Just trying to understand the changes required in general.
> 
> Also, can you confirm if you have tested this patch in your staging env for both w/ and w/o credentials and it all succeeded.
> 
> @codope : I remember you had a local confluent set up right. Do you think you can test this patch out. Also, feel free to review the patch.

Hey @nsivabalan, thanks for reviewing. I can confirm that to get this working we need both the changes to confluent artifacts (upgrading the version) and also the logic changes in the `SchemaRegsitryProvider`. Whilst the logic changes to the `SchemaRegistryProvider` could possibly be added by the user providing their own schema provider it means they have to build and include their own artifact (which would hold their custom schema provider) when they want to use the hudi bundle package. This is huge additional effort for anyone wanting to use the hudi bundle with basic auth for schema registry so i think its quite valuable to include this logic it in the `SchemaRegsitryProvider` class.

I can confirm i have tested this in my staging env with and without credentials and it has all succeeded.

jhsb25 commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/schema/SchemaRegistryProvider.java:
```diff
@@ -48,19 +54,40 @@
+  }
+
+  public InputStream getStream(HttpURLConnection connection) throws IOException {
+    return connection.getInputStream();
```
this is implemented like this for testing purposes, in order to allow for mocking/spying of the InputStream. I've made the method package-private.

jhsb25 commented on hudi-utilities/src/main/java/org/apache/hudi/utilities/schema/SchemaRegistryProvider.java:
```diff
@@ -48,19 +54,40 @@
+    HttpURLConnection connection;
+    Matcher matcher = Pattern.compile("://(.*?)@").matcher(registryUrl);
+    if (matcher.find()) {
+      String creds = matcher.group(1);
```
done

jhsb25 commented:
@nsivabalan all requested changes have been applied, can you re-review and hopefully approve :)

vinothchandar reviewed (approved):
LGTM ! @nsivabalan we can land when you take a final pass. 

@jhsb25 Could you confirm if you have tested these changes on your production setup and life is good?

jhsb25 commented:
> LGTM ! @nsivabalan we can land when you take a final pass.
> 
> @jhsb25 Could you confirm if you have tested these changes on your production setup and life is good?

Thanks for the review @vinothchandar. I can confirm I have tested these changes in my production setup and its all good.

vinothchandar commented on hudi-utilities/src/test/java/org/apache/hudi/utilities/schema/SchemaRegistryProviderTest.java:
```diff
@@ -0,0 +1,111 @@
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+
+class SchemaRegistryProviderTest {
```
My bad for noticing the first time around. can we rename this class to start with - Test.... to be consistent with the rest of the code base

jhsb25 replied:
done

vinothchandar commented on packaging/hudi-integ-test-bundle/pom.xml:
```diff
@@ -615,25 +615,25 @@
       <groupId>io.confluent</groupId>
       <artifactId>common-config</artifactId>
-      <version>3.0.0</version>
+      <version>5.3.4</version>
```
Pull this version into to a property in the root Pom? It will also allow everyone to override it and build against different versions if needed

jhsb25 replied:
done

vinothchandar reviewed (approved):
Just a couple small comments. Again, apologies. Should have caught it before

jhsb25 commented:
> Just a couple small comments. Again, apologies. Should have caught it before

@vinothchandar addressed all the remaining comments now. @nsivabalan can you rereview and approve if possible.

vinothchandar reviewed (approved):
Already approved! Will land once ci is happy.
thanks for the valuable contribution again.
